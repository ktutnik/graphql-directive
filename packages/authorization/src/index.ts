import { MapperKind, mapSchema, SchemaMapper } from "@graphql-tools/utils"
import { getDirectiveValues, GraphQLInputObjectType, GraphQLInputObjectTypeConfig, GraphQLResolveInfo, GraphQLSchema, getNamedType, GraphQLInputFieldConfig, GraphQLArgumentConfig, GraphQLFieldConfigArgumentMap } from "graphql"

// ====================================================== //
// ======================== TYPES ======================= //
// ====================================================== //


interface InvocationContext {
    path: string
    parentValue: any
    resolverArgs: [any, any, any, GraphQLResolveInfo]
}

type Invocable = (value: any, ctx: InvocationContext) => Promise<any>

interface Invocation {
    invoke(value: any, ctx: InvocationContext): Promise<any>
}

class FieldInvocation implements Invocation {
    invoke(value: any, ctx: InvocationContext): Promise<any> {
        throw new Error("Method not implemented.")
    }
}

class TypeInvocation implements Invocation {
    constructor(private name: string) { }

    fields: Record<string, Invocation> = {}

    invoke(value: any, { path, parentValue, ...ctx }: InvocationContext): Promise<unknown> {
        const context = { path: path + "." + this.name, parentValue: value, ...ctx }
        return Promise.all(Object.keys(this.fields).map(key => this.fields[key].invoke(value[key], context)))
    }
}



export { GraphQLSchema }


// ====================================================== //
// ====================== FUNCTIONS ===================== //
// ====================================================== //

type MapperHook<T> = (config: GraphQLInputFieldConfig | GraphQLArgumentConfig, name: string, type: string, sch:GraphQLSchema) => T
type Invoker<T> = (args:GraphQLFieldConfigArgumentMap) => T

function createFieldArgumentInvoker<T>(hook:MapperHook<T>): [SchemaMapper, Invoker<T>] {
    return [{}, ()]
}

const transform = (schema: GraphQLSchema): GraphQLSchema => {
    const [mapper, invoke] = createFieldArgumentInvoker((config) => {

    })

    return mapSchema(schema, {
        ...mapper,
        [MapperKind.OBJECT_FIELD]: (config, fieldName, type, sch) => {
            if (fieldName === "addUser") {
                const result = await invoke(config.args!)
                return config
            }
        },

    })
}

export const createTransformer = () =>
    (schema: GraphQLSchema): GraphQLSchema => {
        return transform(schema)
    }
