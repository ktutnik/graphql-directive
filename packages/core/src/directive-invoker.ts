import { getDirective } from "@graphql-tools/utils"
import { GraphQLArgumentConfig, GraphQLFieldConfigArgumentMap, GraphQLInputFieldConfig, GraphQLInputType, GraphQLResolveInfo, GraphQLSchema, getNamedType, isListType, isNonNullType } from "graphql"

export { GraphQLSchema }

export type ResolverArgs = [any, any, any, GraphQLResolveInfo]

export type InvocationContext = {
    /**
     * List all of directives arguments that is applied in the appropriate field
     */
    directives: Record<string, any>[]

    /**
     * The location of where validator applied from the root path through the GraphQL fields
     */
    path: string

    /**
     * An object shared across all resolvers that are executing for a particular operation. Use this to share per-operation state, including authentication information, dataloader instances, and anything else to track across resolvers.
     */
    contextValue: any

    /**
     * The return value of the resolver for this field's parent (i.e., the previous resolver in the resolver chain). 
     */
    parent: any

    /**
     * An object that contains all GraphQL arguments provided for this field.
     */
    args: any

    /**
     * Contains information about the operation's execution state, including the field name, the path to the field from the root, and more.
     */
    info: GraphQLResolveInfo

}

export type InvokerHook<T> = (value: any, ctx: InvocationContext) => Promise<T[]>


export type DirectiveInvoker<T> = {
    addInputField: (config: GraphQLInputFieldConfig, name: string, type: string, schema: GraphQLSchema) => GraphQLInputFieldConfig
    invoke: (value: any, path:string, argsConfig: GraphQLFieldConfigArgumentMap, resolverArgs: ResolverArgs) => Promise<T[]>
}

type FieldPipeline<T> = {
    name: string,
    proceed: (value: any, path: string, resolverArgs: ResolverArgs) => Promise<T[]>
}

type TypePipeLine<T> = FieldPipeline<T> & { children: FieldPipeline<T>[] }

const createTypePipeline = <T>(name: string): TypePipeLine<T> => {
    const children: TypePipeLine<T>[] = []
    return {
        name,
        children,
        proceed: async (value, path, resolverArgs) => {
            const result = await Promise.all(children.map(child => {
                return child.proceed(value[child.name], path + "." + child.name, resolverArgs)
            }))
            return result.flat()
        }
    }
}

const createChildPipeline = <T>(name: string, next: TypePipeLine<T> | [TypePipeLine<T>] | undefined, hook: InvokerHook<T>, directives: Record<string, any>[]): FieldPipeline<T> => {

    const proceed = async (next: TypePipeLine<T> | undefined, value: any, path: string, resolverArgs: ResolverArgs) => {
        if (value === null || value === undefined) return []
        const nextResult = !!next ? await next.proceed(value, path, resolverArgs) : []
        const [parent, args, contextValue, info] = resolverArgs
        const hookResult = directives.length > 0 ? await hook(value, { directives, parent, args, contextValue, info, path }) : []
        return [...nextResult, ...hookResult]
    }

    return {
        name,
        proceed: async (value, path, resolverArgs) => {
            if (value === null || value === undefined) return []
            if (Array.isArray(next)) {
                const result = await Promise.all((value as any[]).map((val, i) => {
                    return proceed(next[0], val, path + "." + i.toString(), resolverArgs)
                }))
                return result.flat()
            }
            else
                return proceed(next, value, path, resolverArgs)
        }
    }
}



export const createDirectiveInvoker = <T>(directive: string, hook: InvokerHook<T>): DirectiveInvoker<T> => {
    const pipelines: Record<string, TypePipeLine<T>> = {}

    const getPipelineByType = (type: GraphQLInputType): TypePipeLine<T> | [TypePipeLine<T>] | undefined => {
        const ast = getNamedType(type)
        const isList = isNonNullType(type) ? isListType(type.ofType) : isListType(type)
        const pipe = pipelines[ast.name]
        return isList ? [pipe] : pipe
    }

    return {
        addInputField: (config: GraphQLInputFieldConfig | GraphQLArgumentConfig, name: string, type: string, schema: GraphQLSchema) => {
            const directives = getDirective(schema, config, directive) ?? []
            const pipe = pipelines[type] ?? (pipelines[type] = createTypePipeline(type))
            const dataType = getPipelineByType(config.type)
            pipe.children.push(createChildPipeline(name, dataType, hook, directives))
            return config
        },
        invoke: async (value, initPath, configs, resolverArgs) => {
            const result = await Promise.all(Object.keys(configs).map(async (key) => {
                const config = configs[key]
                const dataType = getPipelineByType(config.type)
                const directives = getDirective(resolverArgs[3].schema, config, directive) ?? []
                const pipe = createChildPipeline(key, dataType, hook, directives)
                const path = [initPath, key].filter(Boolean).join(".")
                return pipe.proceed(value[key], path, resolverArgs)
            }))
            return result.flat()
        }
    }
}