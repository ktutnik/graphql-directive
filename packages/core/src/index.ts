export * from "./directive-invoker"
export * from "./validator"