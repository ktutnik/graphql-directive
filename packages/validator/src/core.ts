
export interface TypeConfig {
    kind: "Type"
    name: String
    fields: FieldConfig[]
}

export interface FieldConfig {
    kind: "Field",
    name: string
    validator: Validator
}

export enum ValidationMethod {
    AFTER = "AFTER",
    ALPHA = "ALPHA",
    ALPHANUMERIC = "ALPHANUMERIC",
    ARRAY = "ARRAY",
    ASCII = "ASCII",
    BASE64 = "BASE64",
    BEFORE = "BEFORE",
    BOOLEAN = "BOOLEAN",
    BYTE_LENGTH = "BYTE_LENGTH",
    CREDIT_CARD = "CREDIT_CARD",
    CURRENCY = "CURRENCY",
    DATA_URI = "DATA_URI",
    DECIMAL = "DECIMAL",
    DIVISIBLE_BY = "DIVISIBLE_BY",
    EMAIL = "EMAIL",
    EMPTY = "EMPTY",
    ETHEREUM_ADDRESS = "ETHEREUM_ADDRESS",
    FQDN = "FQDN",
    FLOAT = "FLOAT",
    FULL_WIDTH = "FULL_WIDTH",
    HALF_WIDTH = "HALF_WIDTH",
    HEX_COLOR = "HEX_COLOR",
    HEX_DECIMAL = "HEX_DECIMAL",
    IP = "IP",
    IP_RANGE = "IP_RANGE",
    ISBN = "ISBN",
    ISIN = "ISIN",
    ISO8601 = "ISO8601",
    ISO31661_ALPHA2 = "ISO31661_ALPHA2",
    ISO31661_ALPHA3 = "ISO31661_ALPHA3",
    ISRC = "ISRC",
    ISSN = "ISSN",
    JSON = "JSON",
    JWT = "JWT",
    LAT_LONG = "LAT_LONG",
    LENGTH = "LENGTH",
    LOWERCASE = "LOWERCASE",
    MAC_ADDRESS = "MAC_ADDRESS",
    MIME_TYPE = "MIME_TYPE",
    MONGO_ID = "MONGO_ID",
    MULTIBYTE = "MULTIBYTE",
    NUMBER = "NUMBER",
    PORT = "PORT",
    POSTAL_CODE = "POSTAL_CODE",
    REGEX = "REGEX",
    SURROGATE_PAIR = "SURROGATE_PAIR",
    UPPERCASE = "UPPERCASE",
    URL = "URL",
    UUID = "UUID",
    VARIABLE_WIDTH = "VARIABLE_WIDTH",
    WHITELISTED = "WHITELISTED",
}

export interface DirectiveArgs {
    method: ValidationMethod,
    min: number,
    max: number
}

export interface ErrorMessage {
    path: string,
    message: string
}

export type NativeValidator = (val: any) => string | true

export type Validator = (val: any) => ErrorMessage[] | true