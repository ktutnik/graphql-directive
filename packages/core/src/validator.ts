import { MapperKind, mapSchema } from "@graphql-tools/utils"
import { GraphQLError, GraphQLSchema, defaultFieldResolver } from "graphql"
import { Path } from "graphql/jsutils/Path"
import { InvocationContext, createDirectiveInvokerPipeline } from "./directive-invoker"

// ====================================================== //
// ======================== TYPES ======================= //
// ====================================================== //

const errorCode = "GRAPHQL_VALIDATION_FAILED"

export type Validator = (val: any, ctx: ValidatorContext) => (string | true) | Promise<(string | true)>

export interface ErrorMessage {
    path: string,
    message: string,
}

export interface Plugins {
    [key: string]: Validator
}

export interface TransformerOptions {
    /**
     * List of plugins
     */
    plugins: Plugins,

    /**
     * Name of the directive
     */
    directive: string

    /**
     * List of custom validators
     */
    customValidators?: Plugins
}

export type ValidatorContext = Omit<InvocationContext, "directives"> & {
    /**
     * Contains options values of transformer
     */
    options: TransformerOptions,

    /**
     * Contains argument passed by the @validate directive. For example `@validate(method: LENGTH, min: 1, max: 150)`, the `args` parameter will contains `{ method: LENGTH, min: 1, max: 150 }`.
     */
    directiveArgs: any

}


// ====================================================== //
// ====================== FUNCTIONS ===================== //
// ====================================================== //

const transform = (schema: GraphQLSchema, options: TransformerOptions): GraphQLSchema => {
    const getPath = (path: Path | undefined): string => !!path ? [getPath(path.prev), path.key].filter(Boolean).join(".") : ""
    
    const invoker = createDirectiveInvokerPipeline("validate", async (value, { directives, ...ctx}) => {
        const result = await Promise.all(directives.map(directive => {
            return options.plugins[directive.method](value, { ...ctx, directiveArgs: directive, options })
        }))
        return result.filter((x): x is string => typeof x !== "boolean")
            .map(message => ({ message, path: ctx.path }))
    })

    return mapSchema(schema, {
        [MapperKind.INPUT_OBJECT_FIELD]: invoker.addInputField,
        [MapperKind.OBJECT_FIELD]: (config) => {
            const { resolve = defaultFieldResolver } = config
            return {
                ...config,
                resolve: async (parent, args, context, info) => {
                    const path = getPath(info.path)
                    const error = await invoker.invoke(args, path, config.args!, [parent, args, context, info])
                    if (error.length > 0) {
                        throw new GraphQLError(`Invalid value provided at ${path}`, { extensions: { code: errorCode, error } })
                    }
                    return resolve(parent, args, context, info)
                }
            }
        }
    })
}

export const createValidatorTransformer = (option: TransformerOptions) =>
    (schema: GraphQLSchema, opt: { customValidators?: Plugins } = {}): GraphQLSchema => {
        const customValidatorsPlugin: Plugins = {
            CUSTOM: (val, ctx) => opt.customValidators![ctx.directiveArgs.validator](val, ctx)
        }
        const plugins: Plugins = { ...option.plugins, ...customValidatorsPlugin }
        return transform(schema, { ...option, ...opt, plugins })
    }
