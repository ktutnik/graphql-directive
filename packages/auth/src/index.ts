import { InvocationContext, InvokerHook, createDirectiveInvokerPipeline, createDirectiveInvoker } from "@graphql-directive/core"
import { MapperKind, mapSchema, getDirective } from "@graphql-tools/utils"
import { GraphQLError, GraphQLSchema, OperationTypeNode, defaultFieldResolver, isNonNullType } from "graphql"
import { Path } from "graphql/jsutils/Path"

const typeDefs = /* GraphQL */ `
    directive @authorize(
        policy: String!
    ) on INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION | FIELD_DEFINITION
`

export type AuthorizeContext = Omit<InvocationContext, "directives"> & { directiveArgs: any }

export type PolicyFunction = (ctx: AuthorizeContext) => boolean | Promise<boolean>

export type AuthorizeOptions = {
    policies: Record<string, PolicyFunction>
    queryResolution: "ThrowError" | "Filter"
}

const transform = (schema: GraphQLSchema, options: AuthorizeOptions): GraphQLSchema => {
    const directiveName = "authorize"
    const getPath = (path: Path | undefined): string => !!path ? [getPath(path.prev), path.key].filter(Boolean).join(".") : ""

    const hook: InvokerHook<string> = async (value, { directives, ...ctx }) => {
        const policies = (directives[0].policy as string).split(",").map(x => x.trim())
        const result = await Promise.all(policies.map(name => {
            const policy = options.policies[name]
            if (!policy) throw Error(`Unknown policy "${name}" on ${ctx.path}`)
            return policy({ ...ctx, directiveArgs: policy })
        }))
        return result.some(v => v) ? [] : [ctx.path]
    }

    const pipeline = createDirectiveInvokerPipeline(directiveName, hook)

    return mapSchema(schema, {
        [MapperKind.INPUT_OBJECT_FIELD]: pipeline.addInputField,
        [MapperKind.OBJECT_FIELD]: (config, name, type, schema) => {
            const isNotNull = isNonNullType(config.type)
            const directives = getDirective(schema, config, directiveName) ?? []
            if (directives.length > 0 && options.queryResolution === "Filter" && type !== "Mutation" && isNotNull)
                throw new Error(`Nullable data type is required on ${type}.${name} when Filter resolution enable`)
            const invoker = createDirectiveInvoker(name, hook, directives, undefined)
            const { resolve = defaultFieldResolver } = config
            return {
                ...config,
                resolve: async (parent, args, context, info) => {
                    const operation = info.operation.operation
                    const path = getPath(info.path)
                    const [inputError, fieldError] = await Promise.all([
                        pipeline.invoke(args, path, config.args!, [parent, args, context, info]),
                        invoker.invoke(args, path, [parent, args, context, info])
                    ]);
                    if (inputError.length > 0) {
                        throw new GraphQLError("AUTHORIZATION_ERROR", { extensions: { paths: inputError } })
                    }
                    if (fieldError.length === 0) {
                        return resolve(parent, args, context, info)
                    }
                    if (operation === OperationTypeNode.MUTATION) {
                        throw new GraphQLError("AUTHORIZATION_ERROR", { extensions: { paths: fieldError } })
                    }
                    if (options.queryResolution === "ThrowError") {
                        throw new GraphQLError("AUTHORIZATION_ERROR", { extensions: { paths: fieldError } })
                    }
                    return undefined
                }
            }
        }
    })
}

const createTransformer = (options: Partial<AuthorizeOptions>) => {
    return (schema: GraphQLSchema) => transform(schema, { ... { policies: {}, queryResolution: "Filter" }, ...options })
}

export default {
    typeDefs, createTransformer, 
}