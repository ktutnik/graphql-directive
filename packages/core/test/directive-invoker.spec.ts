import { makeExecutableSchema } from "@graphql-tools/schema"
import { MapperKind, mapSchema } from "@graphql-tools/utils"
import { GraphQLError, graphql } from "graphql"
import { GraphQLSchema, createDirectiveInvoker } from "../src"

const typeDefs = /* GraphQL */ `
    enum ValidationMethod {
        EMAIL, LENGTH, CUSTOM
    }
    directive @test(
        min:Int,
        max:Int
    ) repeatable on INPUT_FIELD_DEFINITION | ARGUMENT_DEFINITION
`

const transform = (schema: GraphQLSchema): GraphQLSchema => {
    const invoker = createDirectiveInvoker("test", async (value, ctx) => {
        return [{ path: ctx.path, directives: ctx.directives }]
    })

    return mapSchema(schema, {
        "MapperKind.INPUT_OBJECT_FIELD": invoker.addInputField,
        [MapperKind.OBJECT_FIELD]: (config, fieldName, type, sch) => {
            if (fieldName === "addUser") {
                return {
                    ...config, resolve: async (source, args, context, info) => {
                        const error = await invoker.invoke(args, "", config.args!, [source, args, context, info])
                        throw new GraphQLError("USER_ERROR", { extensions: { error } })
                    }
                }
            }
        },
    })
}

const val = { typeDefs, transform }


describe("Primitive Type", () => {

    it("Should call directive with primitive type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:String! @test(min: 1, max: 200)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email: "ipsum@dolor.com") }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with array of primitive type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:[String] @test(min: 1, max: 200)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email: ["ipsum@dolor.com", "ipsum@dolor.com"]) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with non nullable array of primitive type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:[String]! @test(min: 1, max: 200)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email: ["ipsum@dolor.com", "ipsum@dolor.com"]) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with non nullable array of non nullable primitive type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:[String!]! @test(min: 1, max: 200)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email: ["ipsum@dolor.com", "ipsum@dolor.com"]) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should able to set multiple directives", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:String! @test(min: 1, max: 200) @test(min: 20)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email: "ipsum@dolor.com") }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should should not calling directive when value is empty", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            type Mutation { 
                addUser(email:String @test(min: 1, max: 200)):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(email:null) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })
})

describe("Custom Type", () => {

    it("Should call directive with custom type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200)
                dob:String!
            }
            
            type Mutation { 
                addUser(user:User!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(user: { name: "Lorem", email: "ipsum@dolor.com", dob: "2023-1-1" }) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with array of custom type properly", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200)
                dob:String!
            }
            
            type Mutation { 
                addUser(user:[User]!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `
            mutation { 
                addUser(user: [
                    { name: "Lorem", email: "ipsum@dolor.com", dob: "2023-1-1" }, 
                    { name: "Lorem", email: "ipsum@dolor.com", dob: "2023-1-1" }
                ]) 
            }`})
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with multiple directive", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200) @test(min: 20)
                dob:String!
            }
            
            type Mutation { 
                addUser(user:User!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({ schema, source: `mutation { addUser(user: { name: "Lorem", email: "ipsum@dolor.com", dob: "2023-1-1" }) }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with with nested properties", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200)
                dob:String!
                partner: User
            }
            
            type Mutation { 
                addUser(user:User!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({
            schema, source: `
            mutation { 
                addUser(user: { 
                    name: "Lorem", 
                    email: "ipsum@dolor.com", 
                    dob: "2023-1-1", 
                    partner: {
                        name: "Lorem", 
                        email: "ipsum@dolor.com", 
                        dob: "2023-1-1"
                    } 
                }) 
            }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should call directive with with nested array properties", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200)
                dob:String!
                partners: [User]
            }
            
            type Mutation { 
                addUser(user:User!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({
            schema, source: `
            mutation { 
                addUser(user: { 
                    name: "Lorem", 
                    email: "ipsum@dolor.com", 
                    dob: "2023-1-1", 
                    partners: [{
                        name: "Lorem", 
                        email: "ipsum@dolor.com", 
                        dob: "2023-1-1"
                    }, {
                        name: "Lorem", 
                        email: "ipsum@dolor.com", 
                        dob: "2023-1-1"
                    }]
                }) 
            }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })

    it("Should not calling directive if provided null on nested properties", async () => {
        const typeDefs = /* GraphQL */ `
            type Query { name:String! }
    
            input User {
                name:String! 
                email:String! @test(min: 1, max: 200)
                dob:String!
                partner: User
            }
            
            type Mutation { 
                addUser(user:User!):Boolean!
            }
        `
        const schema = val.transform(makeExecutableSchema({
            typeDefs: [val.typeDefs, typeDefs],
            resolvers: {
                Mutation: {
                    addUser: () => true
                }
            }
        }))
        const err = await graphql({
            schema, source: `
            mutation { 
                addUser(user: { 
                    name: "Lorem", 
                    email: "ipsum@dolor.com", 
                    dob: "2023-1-1"
                }) 
            }` })
        expect(err.errors![0].extensions).toMatchSnapshot()
    })
})
