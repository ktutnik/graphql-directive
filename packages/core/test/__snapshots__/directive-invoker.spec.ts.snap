// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Type Should call directive with array of custom type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.0.email",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.1.email",
    },
  ],
}
`;

exports[`Custom Type Should call directive with custom type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.email",
    },
  ],
}
`;

exports[`Custom Type Should call directive with multiple directive 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
        {
          "min": 20,
        },
      ],
      "path": "user.email",
    },
  ],
}
`;

exports[`Custom Type Should call directive with with nested array properties 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.email",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.partners.0.email",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.partners.1.email",
    },
  ],
}
`;

exports[`Custom Type Should call directive with with nested properties 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.email",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.partner.email",
    },
  ],
}
`;

exports[`Custom Type Should not calling directive if provided null on nested properties 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "user.email",
    },
  ],
}
`;

exports[`Primitive Type Should able to set multiple directives 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
        {
          "min": 20,
        },
      ],
      "path": "email",
    },
  ],
}
`;

exports[`Primitive Type Should call directive with array of primitive type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.0",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.1",
    },
  ],
}
`;

exports[`Primitive Type Should call directive with non nullable array of non nullable primitive type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.0",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.1",
    },
  ],
}
`;

exports[`Primitive Type Should call directive with non nullable array of primitive type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.0",
    },
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email.1",
    },
  ],
}
`;

exports[`Primitive Type Should call directive with primitive type properly 1`] = `
{
  "error": [
    {
      "directives": [
        {
          "max": 200,
          "min": 1,
        },
      ],
      "path": "email",
    },
  ],
}
`;

exports[`Primitive Type Should should not calling directive when value is empty 1`] = `
{
  "error": [],
}
`;
